common/src/main/java/org/terracotta/angela/common/tcconfig/holders/TcConfigHolder.java

```diff
@@ -106,76 +107,75 @@ public void writeTcConfigFile(File kitDir, final String tcConfigFilename) {
+        //TODO : create client and send command to get free port -> can't connect ? agent exception!
+        // add into xml the port!!
+        // log it!!!
+        // remove below updatePorts method
 
-          Node managementPortNode = (Node) xPath.evaluate("*[name()='management-port']", server, XPathConstants.NODE);
-          int managementPort = managementPortNode == null ? defaultManagementPort() : Integer.parseInt(managementPortNode.getTextContent());
+        int tsaPort = getOrAllocate("tsa-port", server, i, xPath, portAllocator);
+        int jmxPort = support.apply("jmx-port") ? getOrAllocate("jmx-port", server, i, xPath, portAllocator) : 0;
+        int tsaGroupPort = getOrAllocate("tsa-group-port", server, i, xPath, portAllocator);
+        int managementPort = support.apply("management-port") ? getOrAllocate("management-port", server, i, xPath, portAllocator) : 0;
 
-          String symbolicName = nameNode == null ? hostname + ":" + tsaPort : nameNode.getTextContent();
+        String symbolicName = nameNode == null ? hostname + ":" + tsaPort : nameNode.getTextContent();
 
-          TerracottaServer terracottaServer = TerracottaServer
-              .server(symbolicName, hostname)
-              .tsaPort(tsaPort)
-              .tsaGroupPort(tsaGroupPort)
-              .managementPort(managementPort)
-              .jmxPort(jmxPort);
-          servers.add(terracottaServer);
-        }
-      } catch (Exception e) {
-        throw new RuntimeException("Cannot parse tc-config xml", e);
+        TerracottaServer terracottaServer = TerracottaServer
+            .server(symbolicName, hostname)
+            .tsaPort(tsaPort)
+            .tsaGroupPort(tsaGroupPort)
+            .managementPort(managementPort)
+            .jmxPort(jmxPort);
+        servers.add(terracottaServer);
       }
-      this.servers.addAll(servers);
+    } catch (Exception e) {
+      throw new RuntimeException("Cannot parse tc-config xml", e);
     }
-    return servers;
-  }
-
-  protected int defaultManagementPort() {
-    return 9540;
-  }
-
-  protected int defaultTsaGroupPort() {
-    return 9530;
+    this.servers.addAll(servers);
   }
 
-  protected int defaultJmxPort() {
-    return 9520;
+  protected int getOrAllocate(String portName, Node serverNode, int idx, XPath xPath, PortAllocator portAllocator) throws XPathExpressionException {
+    Node node = (Node) xPath.evaluate("*[name()='" + portName + "']", serverNode, XPathConstants.NODE);
+    if (node != null) {
+      final int port = Integer.parseInt(node.getTextContent());
+      logger.debug("Port: node[" + idx + "] " + portName + "=" + port + " (user-set)");
+      return port;
+    } else {
+      int port = portAllocator.reserve(1).next();
+      logger.debug("Port: node[" + idx + "] " + portName + "=" + port + " (random)");
+      updateServerPort(idx, portName, port);
+      return port;
+    }
```

Comment:
Common method to get or allocate ports when server list is initialized.

Bugfix: rewrites the content XML correctly by using the assigned port number. This wasn't done before so it was possible to have a port configured to a default value of 9410 but the generated XML would contain port numbers such as 9111, etc (generated at another place).